annotation_gui_gcp/Database.py

```diff
@@ -0,0 +1,178 @@
+# we want to have a list of image pairs
+# the thingy scrolls on.
+# The gcps are set in a different structure, explained on the notes
+
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from PIL import Image
+from tqdm import tqdm
+
+from opensfm import features
+import numpy as np
+import os
+from collections import OrderedDict
+import json
+from matplotlib.image import _rgb_to_rgba
+
+
+class Database:
+    def __init__(self, _seqs, path, preload_images=True):
+        self.points = OrderedDict()
+        self.path = path
+        self.seqs = _seqs
+        self.image_cache = {}
+
+        if preload_images:
+            print("Preloading images")
+            images = set(self.seqs[0] + self.seqs[1])
+            for image in tqdm(images):
+                self.get_image(image)
+
+        p_gcp_errors = self.get_path() + '/gcp_reprojections.json'
+        if os.path.exists(p_gcp_errors):
+            self.load_gcp_reprojections(p_gcp_errors)
+        else:
+            self.gcp_reprojections = {}
+
+    def load_gcp_reprojections(self, filename):
+        with open(filename, 'r') as f:
+            d = json.load(f)
+
+        for gcp_id in d:
+            for shot_id in d[gcp_id]:
+                h, w = self.get_image_size(shot_id)
+                reproj = d[gcp_id][shot_id]['reprojection']
+                reproj = features.denormalized_image_coordinates(np.array([reproj]), w, h)[0]
+                d[gcp_id][shot_id]['reprojection'] = reproj
+
+        self.gcp_reprojections = d
+
+    def go_to_image_path(self):
+        img_folder = self.path + "/images/"
+        return img_folder
+
+    def get_path(self):
+        return self.path
+
+    def get_seqs(self):
+        return self.seqs
+
+    def init_points(self, points):
+        for point in points:
+            point_id, observations = point.values()
```

Comment:
This fails when the point dictionary has more than 2 entries. Better to access each field explicitly
```suggestion
            point_id, observations = point['id'], point['observations']
```